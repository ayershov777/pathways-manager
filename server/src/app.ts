import express from "express";
import cors from "cors";
import morgan from "morgan";
import { router as pathwaysRouter } from "./routes/pathways.routes";
import { routeNotFoundMiddleware, errorHandlerMiddleware } from "./middleware/error.middleware";

/**
 * Create and configure Express application
 */
const createApp = () => {
    // Initialize Express app
    const app = express();

    // Apply middleware
    app.use(morgan('dev'));
    app.use(cors());
    app.use(express.json());

    // Mount routes
    app.use('/api/v1/pathways', pathwaysRouter);

    // 404 handler - must come after all routes
    app.use(routeNotFoundMiddleware);

    // Global error handler - must be the last middleware
    app.use(errorHandlerMiddleware);

    return app;
};

// Create the application
const app = createApp();

export default app;