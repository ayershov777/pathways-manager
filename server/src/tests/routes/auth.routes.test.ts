import request from 'supertest';
import app from '../../app';
import { User } from '../../models/user.model';
import jwt from 'jsonwebtoken';

// Mock jwt.sign to return predictable tokens for testing
jest.mock('jsonwebtoken');

describe('Auth Routes', () => {
    beforeEach(async () => {
        // Mock jwt sign to return a predictable token
        (jwt.sign as jest.Mock).mockImplementation((payload) => {
            return `mock-token-for-${payload.id}`;
        });

        // Create a test user for login tests
        await User.create({
            email: 'existing@example.com',
            password: 'password123',
            name: 'Existing User'
        });
    });

    describe('POST /api/v1/auth/signup', () => {
        it('should create a new user and return token', async () => {
            const newUser = {
                email: 'new@example.com',
                password: 'password123',
                name: 'New User'
            };

            const response = await request(app)
                .post('/api/v1/auth/signup')
                .send(newUser)
                .expect(201);

            expect(response.body).toMatchSnapshot();

            // Verify user was created in database
            const user = await User.findOne({ email: newUser.email });
            expect(user).toBeDefined();
            expect(user?.name).toBe(newUser.name);
        });

        it('should return 409 if email already exists', async () => {
            const duplicateUser = {
                email: 'existing@example.com',
                password: 'password123',
                name: 'Duplicate User'
            };

            const response = await request(app)
                .post('/api/v1/auth/signup')
                .send(duplicateUser)
                .expect(409);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 400 if required fields are missing', async () => {
            const invalidUser = {
                email: 'invalid@example.com',
                // Missing password and name
            };

            const response = await request(app)
                .post('/api/v1/auth/signup')
                .send(invalidUser)
                .expect(400);

            expect(response.body).toMatchSnapshot();
        });
    });

    describe('POST /api/v1/auth/login', () => {
        it('should authenticate a user and return token', async () => {
            const loginData = {
                email: 'existing@example.com',
                password: 'password123'
            };

            const response = await request(app)
                .post('/api/v1/auth/login')
                .send(loginData)
                .expect(200);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 401 for non-existent user', async () => {
            const loginData = {
                email: 'nonexistent@example.com',
                password: 'password123'
            };

            const response = await request(app)
                .post('/api/v1/auth/login')
                .send(loginData)
                .expect(401);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 401 for incorrect password', async () => {
            const loginData = {
                email: 'existing@example.com',
                password: 'wrongpassword'
            };

            const response = await request(app)
                .post('/api/v1/auth/login')
                .send(loginData)
                .expect(401);

            expect(response.body).toMatchSnapshot();
        });

        it('should return 400 if required fields are missing', async () => {
            const invalidLogin = {
                // Missing email and password
            };

            const response = await request(app)
                .post('/api/v1/auth/login')
                .send(invalidLogin)
                .expect(400);

            expect(response.body).toMatchSnapshot();
        });
    });
});