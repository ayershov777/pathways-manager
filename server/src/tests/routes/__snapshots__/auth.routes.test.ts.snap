// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Routes POST /api/v1/auth/login should authenticate a user and return token 1`] = `
{
  "token": "mock-token-for-68089c9c39a31ea80814c7a3",
  "user": {
    "__v": 0,
    "_id": "68089c9c39a31ea80814c7a3",
    "createdAt": "2025-04-23T07:54:04.469Z",
    "email": "existing@example.com",
    "name": "Existing User",
    "updatedAt": "2025-04-23T07:54:04.469Z",
  },
}
`;

exports[`Auth Routes POST /api/v1/auth/login should return 400 if required fields are missing 1`] = `
{
  "error": {
    "message": "Validation error: email: must have required property 'email', password: must have required property 'password'",
  },
  "success": false,
}
`;

exports[`Auth Routes POST /api/v1/auth/login should return 401 for incorrect password 1`] = `
{
  "error": {
    "message": "Invalid credentials",
  },
  "success": false,
}
`;

exports[`Auth Routes POST /api/v1/auth/login should return 401 for non-existent user 1`] = `
{
  "error": {
    "message": "Invalid credentials",
  },
  "success": false,
}
`;

exports[`Auth Routes POST /api/v1/auth/signup should create a new user and return token 1`] = `
{
  "token": "mock-token-for-68089c9c39a31ea80814c795",
  "user": {
    "__v": 0,
    "_id": "68089c9c39a31ea80814c795",
    "createdAt": "2025-04-23T07:54:04.253Z",
    "email": "new@example.com",
    "name": "New User",
    "updatedAt": "2025-04-23T07:54:04.253Z",
  },
}
`;

exports[`Auth Routes POST /api/v1/auth/signup should return 400 if required fields are missing 1`] = `
{
  "error": {
    "message": "Validation error: password: must have required property 'password', name: must have required property 'name'",
  },
  "success": false,
}
`;

exports[`Auth Routes POST /api/v1/auth/signup should return 409 if email already exists 1`] = `
{
  "error": {
    "message": "Email already in use",
  },
  "success": false,
}
`;
